src/Microsoft.DotNet.Darc/DarcLib/VirtualMonoRepo/CloneManager.cs

```diff
@@ -0,0 +1,160 @@
+
+    public CloneManager(
+        IVmrInfo vmrInfo,
+        IGitRepoCloner gitRepoCloner,
+        ILocalGitClient localGitRepo,
+        ILocalGitRepoFactory localGitRepoFactory,
+        ITelemetryRecorder telemetryRecorder,
+        IFileSystem fileSystem,
+        ILogger<VmrPatchHandler> logger)
+    {
+        _vmrInfo = vmrInfo;
+        _gitRepoCloner = gitRepoCloner;
+        _localGitRepo = localGitRepo;
+        _localGitRepoFactory = localGitRepoFactory;
+        _telemetryRecorder = telemetryRecorder;
+        _fileSystem = fileSystem;
+        _logger = logger;
+    }
+
+    protected async Task<ILocalGitRepo> PrepareCloneInternalAsync(
+        string dirName,
+        IReadOnlyCollection<string> remoteUris,
+        IReadOnlyCollection<string> requestedRefs,
+        string checkoutRef,
+        CancellationToken cancellationToken)
+    {
+        if (remoteUris.Count == 0)
+        {
+            throw new ArgumentException("No remote URIs provided to clone");
+        }
+
+        // Rule out the null commit
+        var refsToVerify = new HashSet<string>(requestedRefs.Where(sha => !Constants.EmptyGitObject.StartsWith(sha)));
+
+        _logger.LogDebug("Fetching refs {refs} from {uris}",
+            string.Join(", ", requestedRefs),
+            string.Join(", ", remoteUris));
+
+        NativePath path = null!;
+        foreach (string remoteUri in remoteUris)
+        {
+            // Path should be returned the same for all invocations
+            // We checkout a default ref
+            path = await PrepareCloneInternal(remoteUri, dirName, cancellationToken);
+            var missingCommit = false;
+
+            // Verify that all requested commits are available
+            foreach (string commit in refsToVerify.ToArray())
+            {
+                try
+                {
+                    var objectType = await _localGitRepo.GetObjectTypeAsync(path, commit);
+                    if (objectType == GitObjectType.Commit)
+                    {
+                        refsToVerify.Remove(commit);
+                    }
+                }
+                catch
+                {
+                    // Ref not found yet, let's try another remote
+                    missingCommit = true;
+                    break;
+                }
+            }
+
+            if (!missingCommit)
+            {
+                _logger.LogDebug("All requested refs ({refs}) found in {repo}", string.Join(", ", requestedRefs), path);
+                break;
+            }
+        }
+
+        if (refsToVerify.Count > 0)
+        {
+            throw new NotFoundException($"Failed to find all requested refs ({string.Join(", ", requestedRefs)}) in {string.Join(", ", remoteUris)}");
+        }
+
+        cancellationToken.ThrowIfCancellationRequested();
+
+        var repo = _localGitRepoFactory.Create(path);
+        await repo.CheckoutAsync(checkoutRef);
+        return repo;
+    }
+
```

Comment:
maybe you can add a comment about what this does, also for PrepareCloneInternalAsync

Reply:
Added, though the public API is already commented with the interfaces